File: src/com/cyanogenmod/setupwizard/wizardmanager/WizardAction.java
/*
 * Copyright (C) 2016 The CyanogenMod Project
 * Copyright (C) 2017 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.cyanogenmod.setupwizard.wizardmanager;

import static android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION;

import static com.cyanogenmod.setupwizard.SetupWizardApp.LOGV;

import android.content.Intent;
import android.os.Parcel;
import android.os.Parcelable;
import android.util.Log;

import com.android.internal.util.XmlUtils;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.net.URISyntaxException;

public class WizardAction implements Parcelable {

    private static final String TAG = "WizardAction";

    private final String mId;
    private final String mUri;
    private final WizardTransitions mTransitions;

    public WizardAction(String id, String uri, WizardTransitions transitions) {
        if (transitions == null) {
            throw new IllegalArgumentException("WizardTransitions cannot be null");
        }
        mId = id;
        mUri = uri;
        mTransitions = transitions;
    }

    public String getId() {
        return mId;
    }

    public String getUri() {
        return mUri;
    }

    public Intent getIntent() {
        Intent intent = null;
        try {
            intent = Intent.parseUri(mUri, FLAG_GRANT_READ_URI_PERMISSION);
        } catch(URISyntaxException e) {
            Log.e(TAG, "Bad URI: " + mUri);
        }
        return intent;
    }

    public String getNextAction(int resultCode) {
        return mTransitions.getAction(resultCode);
    }

    @Override
    public String toString() {
        return "WizardAction{" +
                "mId='" + mId + '\'' +
                ", mUri='" + mUri + '\'' +
                ", mTransitions=" + mTransitions +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof WizardAction)) return false;

        WizardAction that = (WizardAction) o;

        if (mId != null ? !mId.equals(that.mId) : that.mId != null) return false;
        if (mUri != null ? !mUri.equals(that.mUri) : that.mUri != null) return false;
        return mTransitions != null ?
                mTransitions.equals(that.mTransitions) :
                that.mTransitions == null;

    }

    @Override
    public int hashCode() {
        int result = mId != null ? mId.hashCode() : 0;
        result = 31 * result + (mUri != null ? mUri.hashCode() : 0);
        result = 31 * result + (mTransitions != null ? mTransitions.hashCode() : 0);
        return result;
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(mUri);
        dest.writeString(mId);
        dest.writeParcelable(mTransitions, flags);
    }

    public static final Creator<WizardAction> CREATOR = new Creator<WizardAction>() {
        public WizardAction createFromParcel(Parcel source) {
            return new WizardAction(source.readString(),
                    source.readString(),
                    source.readParcelable(WizardTransitions.class.getClassLoader()));
        }

        public WizardAction[] newArray(int size) {
            return new WizardAction[size];
        }
    };

    private static void parseResult(WizardTransitions transitions, XmlPullParser parser) {
        String resultCode = parser.getAttributeValue(WizardScript.WIZARD_SCRIPT_NAMESPACE,
                WizardScript.ATTR_RESULT_CODE);
        String action = parser.getAttributeValue(WizardScript.WIZARD_SCRIPT_NAMESPACE,
                WizardScript.ATTR_ACTION);
        if (LOGV) {
            Log.v(TAG, "parseResult{" +
                    "resultCode='" + resultCode + '\'' +
                    ", action=" + action +
                    '}');
        }
        if (resultCode == null) {
            transitions.setDefaultAction(action);
        } else {
            transitions.put(Integer.valueOf(resultCode).intValue(), action);
        }
    }

    public static WizardAction parseWizardAction(XmlPullParser parser) throws XmlPullParserException,
            IOException {
        String id = parser.getAttributeValue(null, WizardScript.ATTR_ID);
        String uri = parser.getAttributeValue(WizardScript.WIZARD_SCRIPT_NAMESPACE,
                WizardScript.ATTR_URI);
        WizardTransitions transitions = new WizardTransitions();
        if (id == null) {
            throw new XmlPullParserException("WizardAction must define an id");
        }

        if (uri == null) {
            throw new XmlPullParserException("WizardAction must define an intent URI");
        }
        if (LOGV) {
            Log.v(TAG, "parseWizardAction{" +
                    "id='" + id + '\'' +
                    ", uri=" + uri +
                    '}');
        }
        int type;
        final int depth = parser.getDepth();
        while (((type = parser.next()) != XmlPullParser.END_TAG ||
                parser.getDepth() > depth) && type != XmlPullParser.END_DOCUMENT) {
            if (!(type == XmlPullParser.END_TAG || type == XmlPullParser.TEXT)) {
                if (parser.getName().equals(WizardScript.TAG_RESULT)) {
                    parseResult(transitions, parser);
                } else {
                    XmlUtils.skipCurrentTag(parser);
                }
            }
        }

        return new WizardAction(id, uri, transitions);
    }
}

File: app/src/main/java/org/lineageos/setupwizard/wizardmanager/WizardAction.kt
/*
 * SPDX-FileCopyrightText: 2016 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.setupwizard.wizardmanager

import android.content.Intent
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import com.android.internal.util.XmlUtils
import org.lineageos.setupwizard.SetupWizardApp.Companion.LOGV
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException

class WizardAction(val id: String, val uri: String, val transitions: WizardTransitions) :
    Parcelable {

    fun getIntent(): Intent? =
        runCatching {
                Intent.parseUri(uri, Intent.URI_INTENT_SCHEME).apply {
                    addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                }
            }
            .getOrElse {
                Log.e(TAG, "Bad URI: $uri")
                null
            }

    fun getNextAction(resultCode: Int): String? = transitions.getAction(resultCode)

    override fun toString(): String =
        "WizardAction{mId='$id', mUri='$uri', mTransitions=$transitions}"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is WizardAction) return false
        return id == other.id && uri == other.uri && transitions == other.transitions
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + uri.hashCode()
        result = 31 * result + transitions.hashCode()
        return result
    }

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(uri)
        dest.writeParcelable(transitions, flags)
    }

    companion object {
        private const val TAG = "WizardAction"

        @JvmField
        val CREATOR: Parcelable.Creator<WizardAction> =
            object : Parcelable.Creator<WizardAction> {
                override fun createFromParcel(source: Parcel): WizardAction {
                    val id = requireNotNull(source.readString()) { "id missing in parcel" }
                    val uri = requireNotNull(source.readString()) { "uri missing in parcel" }
                    val transitions =
                        requireNotNull(
                            source.readParcelable(
                                WizardTransitions::class.java.classLoader,
                                WizardTransitions::class.java,
                            )
                        ) {
                            "transitions missing in parcel"
                        }
                    return WizardAction(id, uri, transitions)
                }

                override fun newArray(size: Int): Array<WizardAction?> = arrayOfNulls(size)
            }

        private fun parseResult(transitions: WizardTransitions, parser: XmlPullParser) {
            val resultCode =
                parser.getAttributeValue(
                    WizardScript.WIZARD_SCRIPT_NAMESPACE,
                    WizardScript.ATTR_RESULT_CODE,
                )
            val action =
                parser.getAttributeValue(
                    WizardScript.WIZARD_SCRIPT_NAMESPACE,
                    WizardScript.ATTR_ACTION,
                )
            if (LOGV) {
                Log.v(TAG, "parseResult{resultCode='$resultCode', action=$action}")
            }
            if (resultCode == null) {
                transitions.setDefaultAction(action)
            } else {
                transitions.put(resultCode.toInt(), action)
            }
        }

        fun parseWizardAction(parser: XmlPullParser): WizardAction {
            val id =
                parser.getAttributeValue(null, WizardScript.ATTR_ID)
                    ?: throw XmlPullParserException("WizardAction must define an id")
            val uri =
                parser.getAttributeValue(
                    WizardScript.WIZARD_SCRIPT_NAMESPACE,
                    WizardScript.ATTR_URI,
                ) ?: throw XmlPullParserException("WizardAction must define an intent URI")

            if (LOGV) {
                Log.v(TAG, "parseWizardAction{id='$id', uri=$uri}")
            }

            val transitions = WizardTransitions()
            val depth = parser.depth
            var type: Int
            while (
                ((parser.next().also { type = it }) != XmlPullParser.END_TAG ||
                    parser.depth > depth) && type != XmlPullParser.END_DOCUMENT
            ) {
                if (type != XmlPullParser.END_TAG && type != XmlPullParser.TEXT) {
                    if (parser.name == WizardScript.TAG_RESULT) {
                        parseResult(transitions, parser)
                    } else {
                        XmlUtils.skipCurrentTag(parser)
                    }
                }
            }

            return WizardAction(id, uri, transitions)
        }
    }
}
